app/controllers/SocialAuthController.scala

```diff
@@ -0,0 +1,49 @@
+package controllers
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+import com.mohiva.play.silhouette.api._
+import com.mohiva.play.silhouette.api.exceptions.ProviderException
+import com.mohiva.play.silhouette.api.repositories.AuthInfoRepository
+import com.mohiva.play.silhouette.impl.providers._
+import javax.inject.Inject
+import play.api.i18n.{ I18nSupport, Messages }
+import play.api.mvc.{ AbstractController, AnyContent, ControllerComponents, Request }
+
+import models.services.UserService
+import utils.auth.DefaultEnv
+
+class SocialAuthController @Inject() (
+    components: ControllerComponents,
+    silhouette: Silhouette[DefaultEnv],
+    userService: UserService,
+    authInfoRepository: AuthInfoRepository,
+    socialProviderRegistry: SocialProviderRegistry
+  )(implicit ex: ExecutionContext)
+  extends AbstractController(components) with I18nSupport with Logger {
+
+  def authenticate(provider: String) = Action.async { implicit request: Request[AnyContent] =>
+    (socialProviderRegistry.get[SocialProvider](provider) match {
+      case Some(p: SocialProvider with CommonSocialProfileBuilder) =>
+        p.authenticate().flatMap {
+          case Left(result) => Future.successful(result)
+          case Right(authInfo) => for {
+            profile <- p.retrieveProfile(authInfo)
+            user <- userService.save(profile)
+            authInfo <- authInfoRepository.save(profile.loginInfo, authInfo)
```

Comment:
How are you able to reassign `authInfo` here and is it being used? If not, should we prefix with / use `_` to be explicit?

Reply:
local scope i think. the for loop is converted to flatMap statements. https://docs.scala-lang.org/tutorials/FAQ/yield.html, but ya `_` sounds good